Entity/Repository/JobRepository.php

```diff
@@ -152,22 +152,9 @@ public function getOrCreateIfNotExists($command, array $args = array(), $queueNa
         return $firstJob;
     }
 
-    public function findStartableJob($queueName, array &$excludedIds = array())
+    public function findStartableJob($queueName)
     {
-        while (null !== $job = $this->findPendingJob($queueName, $excludedIds)) {
-            if ($job->isStartable()) {
-                return $job;
-            }
-
-            $excludedIds[] = $job->getId();
-
-            // We do not want to have non-startable jobs floating around in
-            // cache as they might be changed by another process. So, better
-            // re-fetch them when they are not excluded anymore.
-            $this->_em->detach($job);
-        }
-
-        return null;
+        return $this->findPendingJob($queueName);
```

Comment:
Revert the breaking changes in the fetching part. And making the association extra lazy. Now the functionality is kept as before. Removed the join from the query, that only acquires unnecessary indices in the call. 

The deadlock is because the locking on indices is not atomic, it will acquire one first and wait for another to be leased. And this Sql requires intention exclusive lock. It will lock all the indices needed. So does the update call, that changes the state from pending to running. Both of them requires "primary" index for field `id` and other indices for field "state".

So I am remove some extra indices and changed the query, making sure only one index will be required and it won't be needed by other update query.